datafusion-common/src/dfschema.rs

```diff
@@ -36,16 +36,30 @@ pub type DFSchemaRef = Arc<DFSchema>;
 pub struct DFSchema {
     /// Fields
     fields: Vec<DFField>,
+    /// Additional metadata in form of key value pairs
+    metadata: HashMap<String, String>,
 }
 
 impl DFSchema {
     /// Creates an empty `DFSchema`
     pub fn empty() -> Self {
-        Self { fields: vec![] }
+        Self {
+            fields: vec![],
+            metadata: HashMap::new(),
+        }
     }
 
+    #[deprecated(since = "7.0.0", note = "please use `new_with_metadata` instead")]
     /// Create a new `DFSchema`
     pub fn new(fields: Vec<DFField>) -> Result<Self> {
+        Self::new_with_metadata(fields, HashMap::new())
+    }
+
+    /// Create a new `DFSchema`
+    pub fn new_with_metadata(
```

Comment:
What would you think about a more "builder-like" API here that perhaps could remain backwards compatible. 

Something like:

```rust
/// Sets the metadata on this DFSchema to `metadata`
pub fn with_metadata(mut self, metadata: HashMap<String, String>) -> Self {
...
}
```


And then instead of code like

```rust
        Schema::new_with_metadata(
            df_schema.fields.iter().map(|f| f.field.clone()).collect(),
            df_schema.metadata.clone(),
        )
```

It would look like
```rust
        Schema::new(
            df_schema.fields.iter().map(|f| f.field.clone()).collect(),
        )
        .with_metadata(df_schema.metadata.clone()),
```

Reply:
Builder-style API is good itself, but is not suitable here I think. 

As you can see, schema conversion between Array and datafusion is used in many places, the compiler will output an warning when `new` is marked as deprecated, but when adopt builder-style API, no warning will be made, and it would be very easy to lost metadata when doing nested conversion.

Reply:
IMO, metadata in DFSchema is a new feature in the next release, if users want to use it, they'll check the API doc. So we don't need to depend on the compiler's warning.

Reply:
As you can see in https://github.com/apache/arrow-datafusion/pull/1914/files#diff-c1ef69547042f0c07aa616c9d5d58cbe2a3c5720f7237c948c99012d6cc0024a

Those optimizers will rewrite plan, if any optimizer forget to attach metadata to newly-created plan, then metadata is lost, it's very easy to miss that without compiler's help.